### PR TITLE
Refactor design tokens to use a proper color table

### PR DESCRIPTION
**Pull request by abey79**
### Related

* Part of/related to https://github.com/rerun-io/rerun/issues/3133
* Part of https://github.com/rerun-io/rerun/issues/3058

### What

This PR refactors the design token as follows:

- All colors from `design_tokens.json` are now loaded in a big table at startup.
- Colors are now referred to using the new the `ColorToken`, which is basically an index into this table.
- Removed all of the color aliases stuff. This was only partially used and very cumbersome to update. The `design_token.rs` file is the _de facto_ source of truth of semantic aliasing for colors.

### CODE REVIEW

File: crates/viewer/re_ui/src/color_table.rs [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+/// Use [`crate::DesignTokens::color`] to get the color corresponding to a token.
+#[derive(Debug, Clone, Copy, Hash)]
+pub struct ColorToken {
+    hue: Hue,
+    shade: Shade,
+}
+
+impl ColorToken {
+    #[inline]
+    pub fn new(hue: Hue, shade: Shade) -> Self {
```
Comment (abey79) [inline crates/viewer/re_ui/src/color_table.rs:189]: With `enumn` (which we already depend on) I could easy add support for `u16` input, but then we must deal with error management. For that, we could either:
- panic
- have an invalid state for `ColorToken` (e.g. `shade = None`), which would return a random color when queried (so it flashes in the UI) and logs an error

File: crates/viewer/re_ui/src/color_table.rs [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+            Self::Red => f.write_str("Red"),
+            Self::Blue => f.write_str("Blue"),
+            Self::Purple => f.write_str("Purple"),
+        }
+    }
+}
+
+/// A shade for a [`ColorToken`].
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq)]
+pub enum Shade {
```
Comment (emilk) [inline crates/viewer/re_ui/src/color_table.rs:37]: Please document what `Shade` means (at least roughly). Is `S1000` fully saturated, and `S0` gray? Or is there no such inherent meaning?

File: crates/viewer/re_ui/src/color_table.rs [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+    pub fn as_u16(self) -> u16 {
+        self as u16
+    }
+}
+
+/// A table mapping all combination of [`Hue`] and [`Shade`] to a [`egui::Color32`].
+#[derive(Debug)]
+pub struct ColorTable {
+    pub color_table: nohash_hasher::IntMap<Hue, nohash_hasher::IntMap<Shade, egui::Color32>>,
```
Comment (emilk) [inline crates/viewer/re_ui/src/color_table.rs:139]: You're using `nohash_hasher` here, presumably for speed.

If you really want full speed, use sequential numbers of the enum variants (that's the default) and then use a LUT

File: crates/viewer/re_ui/src/color_table.rs [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+            Shade::S950,
+            Shade::S975,
+            Shade::S1000,
+        ];
+        &ALL
+    }
+
+    pub fn as_u16(self) -> u16 {
+        self as u16
+    }
```
Comment (emilk) [inline crates/viewer/re_ui/src/color_table.rs:133]: What will this be used for?

File: crates/viewer/re_ui/src/design_tokens.rs [hunk 5]
```diff
@@ -418,23 +428,36 @@ impl DesignTokens {
...
+    ColorTable {
+        color_table: Hue::all()
+            .iter()
+            .map(|hue| {
+                let shade_map = Shade::all()
+                    .iter()
+                    .map(|shade| {
+                        let color = get_color_from_json(
+                            json,
+                            &format!("{{Global.Color.{hue}.{}}}", shade.as_u16()),
```
Comment (emilk) [inline crates/viewer/re_ui/src/design_tokens.rs:452]: Just implement `Display` for `Shade` instead? 🤷 

```suggestion
                            &format!("{{Global.Color.{hue}.{shade}}}"),
```

**Review by emilk — APPROVED** [review]
Very nice to finally get this!

File: crates/viewer/re_ui/src/color_table.rs (hunk 3 above)
Comment (abey79) [reply crates/viewer/re_ui/src/color_table.rs:139]: Do you have a suggestion for a pattern/helper crate for obtaining at compile time the number of variant of an enum and a list of all variants?

Comment (emilk) [reply crates/viewer/re_ui/src/color_table.rs:139]: https://doc.rust-lang.org/stable/std/mem/fn.variant_count.html :]

Comment (abey79) [reply crates/viewer/re_ui/src/color_table.rs:139]: that looks really cool! but...

image

Comment (abey79) [reply crates/viewer/re_ui/src/color_table.rs:139]: oh I can use `strum`

**Pull request merged**